File: Scanner-Backend/app/api/webscan/acunetix.py
"""Acunetix Premium (on-prem, v13+) REST client.

Only the *create and start* half of the lifecycle lives here. Once a scan is
running, the Go ``worker-webscan`` polls Acunetix (and pulls the findings) so
the API process never blocks on a scan that can take minutes to hours.

Authentication uses the API key generated in Acunetix →
*Profile → API key*, sent as the ``X-Auth`` header.

The base URL is read from ``ACUNETIX_URL`` (or its older alias
``ACUNETIX_BASE_URL``). Either the bare console address
(``https://acunetix.example.com:3443``) or the API root
(``https://acunetix.example.com:3443/api/v1``) is accepted — see
:func:`resolve_base_url`, which normalizes both to the API root that the paths
below are appended to.
"""

import os
import logging

import httpx

logger = logging.getLogger(__name__)


class AcunetixError(RuntimeError):
    """Raised when Acunetix is misconfigured, unreachable, or answers with an error."""


def _env_bool(name: str, default: bool = False) -> bool:
    raw = os.getenv(name)
    if raw is None or str(raw).strip() == "":
        return default
    return str(raw).strip().lower() in {"1", "true", "yes", "on"}


def _env_float(name: str, default: float) -> float:
    raw = os.getenv(name)
    if raw is None or str(raw).strip() == "":
        return default
    try:
        return float(str(raw).strip())
    except ValueError:
        logger.warning("%s=%r is not a number; using %s", name, raw, default)
        return default


def resolve_base_url(raw: str | None = None) -> str:
    """Resolve and normalize the Acunetix API root.

    Reads ``ACUNETIX_URL`` first, then ``ACUNETIX_BASE_URL``. Accepts the bare
    console address, a trailing slash, or a browser-style fragment
    (``https://host:3443/#/dashboard``) and always returns the ``/api/v1`` root.
    A sub-path installation (``https://host/acunetix/api/v1``) is preserved.

    Returns an empty string when nothing is configured.
    """
    candidate = (
        raw
        or os.getenv("ACUNETIX_URL")
        or os.getenv("ACUNETIX_BASE_URL")
        or ""
    ).strip()
    if not candidate:
        return ""

    # Drop a browser-console fragment or query string (…/index.php#/dashboard).
    candidate = candidate.split("#", 1)[0].split("?", 1)[0].strip()
    if not candidate:
        return ""

    if "://" not in candidate:
        candidate = f"https://{candidate}"

    candidate = candidate.rstrip("/")
    lowered = candidate.lower()

    if lowered.endswith("/api/v1"):
        return candidate
    if lowered.endswith("/api"):
        return f"{candidate}/v1"
    return f"{candidate}/api/v1"


def _error_detail(response: httpx.Response) -> str:
    """Best-effort human-readable error body from an Acunetix response."""
    try:
        payload = response.json()
    except ValueError:
        return (response.text or "").strip()[:500] or response.reason_phrase

    if isinstance(payload, dict):
        for key in ("message", "error", "detail", "Message"):
            value = payload.get(key)
            if value:
                return str(value)[:500]
    return str(payload)[:500]


class AcunetixClient:
    """Thin synchronous wrapper around the Acunetix Premium v1 API."""

    def __init__(
        self,
        base_url: str | None = None,
        api_key: str | None = None,
        verify_tls: bool | None = None,
        timeout: float | None = None,
    ):
        self.base_url = resolve_base_url(base_url)
        self.api_key = (api_key or os.getenv("ACUNETIX_API_KEY") or "").strip()
        if not self.base_url:
            raise AcunetixError(
                "ACUNETIX_URL is not set. Point it at your Acunetix console, "
                "e.g. https://acunetix.example.com:3443"
            )
        if not self.api_key:
            raise AcunetixError(
                "ACUNETIX_API_KEY is not set. Generate an API key in Acunetix "
                "(Profile → API key) and add it to the environment."
            )

        # On-prem installs normally present a self-signed certificate, so TLS
        # verification is off unless explicitly enabled. Every Acunetix setting is
        # read from the environment — no URL, key or limit lives in this module.
        if verify_tls is None:
            verify_tls = _env_bool("ACUNETIX_VERIFY_TLS", False)
        if timeout is None:
            timeout = _env_float("ACUNETIX_HTTP_TIMEOUT_SEC", 30.0)

        self._client = httpx.Client(
            base_url=self.base_url,
            headers={
                "X-Auth": self.api_key,
                "Content-Type": "application/json",
                "Accept": "application/json",
            },
            verify=verify_tls,
            timeout=timeout,
        )

    # ── lifecycle ─────────────────────────────────────────────────────────────

    def __enter__(self) -> "AcunetixClient":
        return self

    def __exit__(self, *_exc) -> None:
        self.close()

    def close(self) -> None:
        try:
            self._client.close()
        except Exception:  # pragma: no cover - defensive
            pass

    # ── internals ─────────────────────────────────────────────────────────────

    def _request(self, method: str, path: str, **kwargs) -> dict:
        try:
            response = self._client.request(method, path, **kwargs)
        except httpx.HTTPError as exc:
            raise AcunetixError(f"Could not reach Acunetix at {self.base_url}{path}: {exc}") from exc

        if response.status_code >= 400:
            raise AcunetixError(
                f"Acunetix {method} {path} returned {response.status_code}: {_error_detail(response)}"
            )

        if not response.content:
            return {}
        try:
            payload = response.json()
        except ValueError:
            return {}
        return payload if isinstance(payload, dict) else {"data": payload}

    # ── targets ───────────────────────────────────────────────────────────────

    def find_target_by_address(self, address: str) -> str | None:
        """Return the id of an existing target with exactly this address, if any.

        Acunetix' ``q`` parameter is a structured filter expression
        (``q=(criticality:equals:high)``), not a free-text search, so passing a
        raw URL makes ``GET /targets`` fail with ``400 Validation errors``.
        Pagination is cursor-based (``l`` limit + ``c`` cursor), so we page
        through the targets and match on the address client-side.
        """
        wanted = (address or "").strip().rstrip("/").lower()
        if not wanted:
            return None

        cursor: str | None = ""
        # Hard cap so a broken pagination cursor can never loop forever.
        for _ in range(200):
            params: dict = {"l": 100}
            if cursor:
                params["c"] = cursor
            payload = self._request("GET", "/targets", params=params)

            for target in payload.get("targets") or []:
                candidate = str(target.get("address") or "").strip().rstrip("/").lower()
                if candidate == wanted:
                    return str(target.get("target_id") or "") or None

            pagination = payload.get("pagination") or {}
            cursor = str(
                pagination.get("cursor") or pagination.get("c") or ""
            ).strip() or None
            if not cursor:
                break
        return None

    def create_target(self, address: str, description: str = "", criticality: int = 10) -> str:
        """Create (or reuse) an Acunetix target and return its ``target_id``."""
        existing = self.find_target_by_address(address)
        if existing:
            logger.info("Reusing existing Acunetix target %s for %s", existing, address)
            return existing

        payload = self._request(
            "POST",
            "/targets",
            json={
                "address": address,
                "description": description or f"ShieldStat web scan for {address}",
                "criticality": criticality,
            },
        )
        target_id = str(payload.get("target_id") or "").strip()
        if not target_id:
            raise AcunetixError("Acunetix did not return a target_id when creating the target.")
        logger.info("Created Acunetix target %s for %s", target_id, address)
        return target_id

    # ── scanning profiles ─────────────────────────────────────────────────────

    def list_profiles(self) -> list[dict]:
        """Return the scanning profiles Acunetix exposes."""
        payload = self._request("GET", "/scanning_profiles")
        return payload.get("scanning_profiles") or []

    def resolve_profile_id(self) -> str:
        """Pick the scanning profile to use.

        ``ACUNETIX_PROFILE_ID`` wins when set. Otherwise a profile whose name
        looks like a full scan is used. When nothing matches, an empty string is
        returned so the caller can omit ``profile_id`` and let Acunetix apply its
        own default (normally *Full Scan*) — blindly using the first returned
        profile can select a crawl-only one that reports no vulnerabilities.
        """
        configured = (os.getenv("ACUNETIX_PROFILE_ID") or "").strip()
        if configured:
            return configured

        profiles = self.list_profiles()
        if not profiles:
            raise AcunetixError(
                "Acunetix reported no scanning profiles. Set ACUNETIX_PROFILE_ID "
                "to the profile you want to use."
            )

        for profile in profiles:
            name = str(profile.get("name") or "").lower()
            if "full" in name:
                return str(profile.get("profile_id") or "").strip()

        available = ", ".join(
            sorted({str(p.get("name") or "").strip() for p in profiles if p.get("name")})
        )
        logger.warning(
            "No Acunetix scanning profile matched a full scan; leaving the profile "
            "unset so Acunetix uses its default. Available profiles: %s. Pin one "
            "with ACUNETIX_PROFILE_ID to be explicit.",
            available,
        )
        return ""

    # ── scans ─────────────────────────────────────────────────────────────────

    def start_scan(self, target_id: str, profile_id: str | None = None) -> str:
        """Start a scan immediately (not scheduled) and return its ``scan_id``.

        ``profile_id`` is omitted from the request when it resolves to empty, so
        Acunetix falls back to its own default profile.
        """
        profile = profile_id if profile_id is not None else self.resolve_profile_id()
        profile = str(profile or "").strip()

        body: dict = {
            "target_id": target_id,
            "schedule": {
                "disable": False,
                "start_date": None,
                "time_sensitive": False,
            },
        }
        if profile:
            body["profile_id"] = profile

        payload = self._request("POST", "/scans", json=body)
        scan_id = str(payload.get("scan_id") or "").strip()
        if not scan_id:
            raise AcunetixError("Acunetix did not return a scan_id when starting the scan.")
        logger.info("Started Acunetix scan %s for target %s", scan_id, target_id)
        return scan_id

    def abort_scan(self, scan_id: str) -> None:
        """Ask Acunetix to stop a running scan (best effort)."""
        try:
            self._request("POST", f"/scans/{scan_id}/abort")
        except AcunetixError:
            logger.warning("Could not abort Acunetix scan %s", scan_id, exc_info=True)

File: scanner-platform/internal/worker/acunetix.go
package worker

import (
	"crypto/tls"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"os"
	"strconv"
	"strings"
	"time"
)

// acunetixClient is a read-only Acunetix Premium (v13+) API client.
//
// The backend already creates the target and starts the scan; this client only
// watches the scan and pulls its findings, so it needs no write endpoints.
type acunetixClient struct {
	baseURL string
	apiKey  string
	http    *http.Client
}

// resolveAcunetixBaseURL resolves and normalizes the Acunetix API root.
//
// Reads ACUNETIX_URL first, then its alias ACUNETIX_BASE_URL. Accepts the bare
// console address, a trailing slash, or a browser-style fragment and always
// returns the /api/v1 root, preserving sub-path installations.
func resolveAcunetixBaseURL() string {
	raw := strings.TrimSpace(os.Getenv("ACUNETIX_URL"))
	if raw == "" {
		raw = strings.TrimSpace(os.Getenv("ACUNETIX_BASE_URL"))
	}
	if raw == "" {
		return ""
	}

	// Drop a browser-console fragment or query string.
	if idx := strings.IndexAny(raw, "#?"); idx >= 0 {
		raw = raw[:idx]
	}
	raw = strings.TrimSpace(raw)
	if raw == "" {
		return ""
	}

	if !strings.Contains(raw, "://") {
		raw = "https://" + raw
	}
	raw = strings.TrimRight(raw, "/")

	lowered := strings.ToLower(raw)
	switch {
	case strings.HasSuffix(lowered, "/api/v1"):
		return raw
	case strings.HasSuffix(lowered, "/api"):
		return raw + "/v1"
	default:
		return raw + "/api/v1"
	}
}

// newAcunetixClientFromEnv builds a client from ACUNETIX_URL (or
// ACUNETIX_BASE_URL) and ACUNETIX_API_KEY.
func newAcunetixClientFromEnv() (*acunetixClient, error) {
	baseURL := resolveAcunetixBaseURL()
	apiKey := strings.TrimSpace(os.Getenv("ACUNETIX_API_KEY"))

	if baseURL == "" {
		return nil, fmt.Errorf("ACUNETIX_URL environment variable is not set")
	}
	if apiKey == "" {
		return nil, fmt.Errorf("ACUNETIX_API_KEY environment variable is not set")
	}

	transport := http.DefaultTransport.(*http.Transport).Clone()
	// On-prem Acunetix installs normally use a self-signed certificate, so TLS
	// verification is off unless ACUNETIX_VERIFY_TLS is explicitly enabled.
	transport.TLSClientConfig = &tls.Config{InsecureSkipVerify: !envBool("ACUNETIX_VERIFY_TLS", false)}

	return &acunetixClient{
		baseURL: baseURL,
		apiKey:  apiKey,
		http: &http.Client{
			Timeout:   time.Duration(envInt("ACUNETIX_HTTP_TIMEOUT_SEC", defaultAcunetixHTTPTimeoutSec)) * time.Second,
			Transport: transport,
		},
	}, nil
}

func (c *acunetixClient) get(path string, query url.Values) (map[string]any, error) {
	endpoint := c.baseURL + path
	if len(query) > 0 {
		endpoint += "?" + query.Encode()
	}

	req, err := http.NewRequest(http.MethodGet, endpoint, nil)
	if err != nil {
		return nil, fmt.Errorf("building request for %s: %w", path, err)
	}
	req.Header.Set("X-Auth", c.apiKey)
	req.Header.Set("Accept", "application/json")

	res, err := c.http.Do(req)
	if err != nil {
		return nil, fmt.Errorf("request to %s failed: %w", path, err)
	}
	defer res.Body.Close()

	// 32 MB cap — a chatty scan can return a large vulnerability list.
	body, err := io.ReadAll(io.LimitReader(res.Body, 32<<20))
	if err != nil {
		return nil, fmt.Errorf("reading response from %s: %w", path, err)
	}

	if res.StatusCode < 200 || res.StatusCode >= 300 {
		return nil, fmt.Errorf("acunetix %s returned %s: %s", path, res.Status, strings.TrimSpace(string(body)))
	}

	if len(body) == 0 {
		return map[string]any{}, nil
	}

	var payload map[string]any
	if err := json.Unmarshal(body, &payload); err != nil {
		return nil, fmt.Errorf("decoding response from %s: %w", path, err)
	}
	return payload, nil
}

// listPaginated walks Acunetix' cursor pagination for a collection endpoint.
func (c *acunetixClient) listPaginated(path, collectionKey string) ([]map[string]any, error) {
	var all []map[string]any
	query := url.Values{"l": {"100"}}

	// Hard cap so a broken cursor can never spin forever.
	for page := 0; page < 200; page++ {
		payload, err := c.get(path, query)
		if err != nil {
			return nil, err
		}

		items := asMapSlice(payload[collectionKey])
		all = append(all, items...)

		cursor := ""
		if pagination, ok := payload["pagination"].(map[string]any); ok {
			cursor = stringField(pagination, "cursor")
		}
		if cursor == "" || len(items) == 0 {
			break
		}
		query.Set("c", cursor)
	}

	return all, nil
}

func (c *acunetixClient) getScan(scanID string) (map[string]any, error) {
	return c.get("/scans/"+url.PathEscape(scanID), nil)
}

func (c *acunetixClient) listScanResults(scanID string) ([]map[string]any, error) {
	return c.listPaginated("/scans/"+url.PathEscape(scanID)+"/results", "results")
}

func (c *acunetixClient) listVulnerabilities(scanID, resultID string) ([]map[string]any, error) {
	path := "/scans/" + url.PathEscape(scanID) + "/results/" + url.PathEscape(resultID) + "/vulnerabilities"

	vulns, err := c.listPaginated(path, "vulnerabilities")
	if err != nil || len(vulns) > 0 {
		return vulns, err
	}

	// Older builds return the collection under "vulns".
	return c.listPaginated(path, "vulns")
}

// getVulnerability fetches one vulnerability's detail (affected URL, evidence).
func (c *acunetixClient) getVulnerability(scanID, resultID, vulnID string) (map[string]any, error) {
	path := "/scans/" + url.PathEscape(scanID) +
		"/results/" + url.PathEscape(resultID) +
		"/vulnerabilities/" + url.PathEscape(vulnID)
	return c.get(path, nil)
}

// getVulnerabilityType returns the check's metadata (name, CVSS, CWE, remedy).
func (c *acunetixClient) getVulnerabilityType(vtID string) (map[string]any, error) {
	return c.get("/vulnerability_types/"+url.PathEscape(vtID), nil)
}

// ─── JSON helpers ─────────────────────────────────────────────────────────────

func asMapSlice(value any) []map[string]any {
	raw, ok := value.([]any)
	if !ok {
		return nil
	}
	out := make([]map[string]any, 0, len(raw))
	for _, item := range raw {
		if mapped, ok := item.(map[string]any); ok {
			out = append(out, mapped)
		}
	}
	return out
}

func stringField(payload map[string]any, key string) string {
	if payload == nil {
		return ""
	}
	value, ok := payload[key]
	if !ok || value == nil {
		return ""
	}
	if str, ok := value.(string); ok {
		return strings.TrimSpace(str)
	}
	return strings.TrimSpace(fmt.Sprintf("%v", value))
}

func intField(payload map[string]any, key string) int {
	if payload == nil {
		return 0
	}
	value, _ := toInt(payload[key])
	return value
}

func toInt(value any) (int, bool) {
	switch typed := value.(type) {
	case float64:
		return int(typed), true
	case int:
		return typed, true
	case int64:
		return int(typed), true
	case string:
		parsed, err := strconv.Atoi(strings.TrimSpace(typed))
		if err != nil {
			return 0, false
		}
		return parsed, true
	default:
		return 0, false
	}
}

func stringSlice(value any) []string {
	if value == nil {
		return nil
	}
	if str, ok := value.(string); ok {
		if strings.TrimSpace(str) == "" {
			return nil
		}
		return []string{strings.TrimSpace(str)}
	}

	raw, ok := value.([]any)
	if !ok {
		return nil
	}
	out := make([]string, 0, len(raw))
	for _, item := range raw {
		if str, ok := item.(string); ok && strings.TrimSpace(str) != "" {
			out = append(out, strings.TrimSpace(str))
		}
	}
	return out
}

func firstNonEmpty(values ...string) string {
	for _, value := range values {
		if strings.TrimSpace(value) != "" {
			return strings.TrimSpace(value)
		}
	}
	return ""
}

func envBool(name string, fallback bool) bool {
	raw := strings.ToLower(strings.TrimSpace(os.Getenv(name)))
	if raw == "" {
		return fallback
	}
	switch raw {
	case "1", "true", "yes", "on":
		return true
	default:
		return false
	}
}

func envInt(name string, fallback int) int {
	raw := strings.TrimSpace(os.Getenv(name))
	if raw == "" {
		return fallback
	}
	parsed, err := strconv.Atoi(raw)
	if err != nil || parsed <= 0 {
		return fallback
	}
	return parsed
}
